runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/ProjectionPushdownOptimizerTest.java

```diff
@@ -0,0 +1,364 @@
+import org.apache.beam.sdk.values.TupleTagList;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterables;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link ProjectionPushdownOptimizer}. */
+@RunWith(JUnit4.class)
+public class ProjectionPushdownOptimizerTest {
+
+  @Test
+  public void testSourceDoesNotImplementPushdownProjector() {
+    Pipeline p = Pipeline.create();
+    SimpleSource source = new SimpleSource(FieldAccessDescriptor.withAllFields());
+    p.apply(source)
+        .apply(new FieldAccessTransform(FieldAccessDescriptor.withFieldNames("foo", "bar")));
+
+    ProjectionPushdownOptimizer.optimize(p);
+    Assert.assertTrue(pipelineHasTransform(p, source));
+  }
+
+  @Test
+  public void testSimpleProjectionPushdown() {
+    Pipeline p = Pipeline.create();
+    SimpleSourceWithPushdown originalSource =
+        new SimpleSourceWithPushdown(FieldAccessDescriptor.withAllFields());
+    FieldAccessDescriptor downstreamFieldAccess =
+        FieldAccessDescriptor.withFieldNames("foo", "bar");
+    p.apply(originalSource).apply(new FieldAccessTransform(downstreamFieldAccess));
+
+    SimpleSourceWithPushdown expectedSource = new SimpleSourceWithPushdown(downstreamFieldAccess);
+
+    ProjectionPushdownOptimizer.optimize(p);
+    Assert.assertTrue(pipelineHasTransform(p, expectedSource));
+    Assert.assertFalse(pipelineHasTransform(p, originalSource));
+  }
+
+  @Test
+  public void testBranchedProjectionPushdown() {
+    Pipeline p = Pipeline.create();
+    SimpleSourceWithPushdown originalSource =
+        new SimpleSourceWithPushdown(FieldAccessDescriptor.withAllFields());
+    PCollection<Row> input = p.apply(originalSource);
+    input.apply(new FieldAccessTransform(FieldAccessDescriptor.withFieldNames("foo")));
+    input.apply(new FieldAccessTransform(FieldAccessDescriptor.withFieldNames("bar")));
+
+    SimpleSourceWithPushdown expectedSource =
+        new SimpleSourceWithPushdown(FieldAccessDescriptor.withFieldNames("foo", "bar"));
+
+    ProjectionPushdownOptimizer.optimize(p);
+    Assert.assertTrue(pipelineHasTransform(p, expectedSource));
+    Assert.assertFalse(pipelineHasTransform(p, originalSource));
+  }
+
+  @Test
+  public void testIntermediateProducer() {
+    Pipeline p = Pipeline.create();
+    SimpleSource source = new SimpleSource(FieldAccessDescriptor.withAllFields());
+    IntermediateTransformWithPushdown originalT =
+        new IntermediateTransformWithPushdown(FieldAccessDescriptor.withAllFields());
+    FieldAccessDescriptor downstreamFieldAccess =
+        FieldAccessDescriptor.withFieldNames("foo", "bar");
+    p.apply(source).apply(originalT).apply(new FieldAccessTransform(downstreamFieldAccess));
+
+    // TODO(BEAM-13658) Support pushdown on intermediate transforms.
+    // For now, test that the pushdown optimizer ignores immediate transforms.
```

Comment:
I wanted to make sure the optimizer wouldn't throw an exception in this case, since that would break the user's pipeline (and it was throwing before I added this test case and fixed the issue).